tests/test_checkout.py

```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
         r2_checkout.close()
         w_checkout.close()
 
     def test_write_with_r_checkout(self, written_repo, array5by7):
         co = written_repo.checkout()
+        # TODO: This should raise more informative error message
+        # Now it's type error because `init_dataset` is None
         with pytest.raises(TypeError):
             co.datasets.init_dataset(name='dset', shape=(5, 7), dtype=np.float64)
-            co.datasets['_dset']['1'] = array5by7
-            co.datasets['_dset'].add('1', array5by7)
+        # TODO: This should be permission error
+        with pytest.raises(AttributeError):
+            co.metadata.add('a', 'b')
         co.close()
 
     def test_write_after_repo_deletion(self, written_repo, array5by7):
         co = written_repo.checkout(write=True)
         shutil.rmtree(written_repo._repo_path)
         with pytest.raises(OSError):
             co.datasets.init_dataset('dset', prototype=array5by7)
+        co.metadata.add('a', 'b')
         dset = co.datasets['_dset']
         dset['1'] = array5by7
-        # todo: shouldn't it be atleast same as above
         with pytest.raises(FileNotFoundError):
             co.commit()
         co.close()
 
-
-@pytest.mark.skip(reason='not implemented')
-class TestBranching(object):
-
-    def test_merge(self):
-        pass
-
-    def test_name_conflict(self):
-        pass
-
-    def test_delete_branch(self):
-        pass
-
-    def test_merge_multiple_checkouts(self):
-        pass
-
-    def test_merge_diverged(self):
-        pass
-
-    def test_merge_diverged_conflict(self):
-        pass
+    def test_operate_on_dataset_after_closing_old_co(self, repo, array5by7):
+        co = repo.checkout(write=True)
+        dset = co.datasets.init_dataset('dset', prototype=array5by7)
+        co.commit()
+        co.close()
+        co = repo.checkout(write=True)
+        # TODO: what is the design here. IMO, the old instance should not execute
+        dset.add(array5by7, '1')
+        co.commit()
+        co.close()
+        # TODO: even after the close, I could still fetch the
+        #   data from the old dset object.
+        print(dset['1'])
+
+    def test_operate_on_closed_checkout(self, repo, array5by7):
+        co = repo.checkout(write=True)
+        co.datasets.init_dataset('dset', prototype=array5by7)
+        co.commit()
+        co.close()
+        # TODO: This should raise permission error or something
+        with pytest.raises(KeyError):
+            co.datasets['dset']['1'] = array5by7
+        co.metadata.add('a', 'b')
+
+
+class TestBranching:
+
+    def test_invalid_name(self, repo):
+        with pytest.raises(ValueError):
+            repo.create_branch('dummy branch')
+
+    def test_new_branch_from_new_repo(self, repo):
+        branch = repo.create_branch('testbranch')
+        # TODO: it doesnt raise value error to the user
+        # instead return false and print out the value error
+        # Ideally, it should let users create new branch
```

Comment:
Nope. For now, I want users to explicitly say they want to create a branch. 

To better explain the reasoning here: I'm hard set on being explicit for branch operations because the model is slightly different from `Git`'s in the sense that there is no way to do a `write-enabled checkout` on a `detached HEAD commit`. All writes have to be performed on the `"tip" HEAD commit` of a branch. Without being explicit about the branch creation and checkout operations, this subtle difference might lead to confusion in day-to-day use.

Reply:
Correct me If I am wrong; What I understand from the context above is, on a new repository user should not be able to create a new branch until the first commit?

Reply:
@rlizzo I am good to merge the PR once this conversation can be resolved (either through a Fix or an issue that tracks this test case). Let me know your thoughts. As a first-time hangar user, I would expect hangar to let me create a branch on an empty repository